### PR TITLE
Add constantine

### LINKED ISSUE #13: Add Constantine as Git Submodule
Add the ***constantine library*** as git submodule to this repository
The constantine lib -> [https://github.com/mratsim/constantine/](https://github.com/mratsim/constantine/)

This will provide us the following for the verkle tree : 
 - Types for elliptic curve points
 - Banderwagon Scheme over the Bandersnatch Curve
 - Serialisation & Deserialisation functions
 - Inner Product Arguments & Pederson Commits 
 - Multiproofs 

In Future:
 - Make constantine a nimble installable

### PR DESCRIPTION
**Pull request by advaita-saha**
Fixes #13

### CODE REVIEW

**Comment by mratsim** [comment]
I can create a tag or you can target a specific commit, just in case I somehow break something.

**Comment by agnxsh** [comment]
> I can create a tag or you can target a specific commit, just in case I somehow break something.

Also, our plan is, we can work with the updated master branch adding it as a submodule and start adding the types, my PR in Constantine is taking me some time to test and debug, so I'm planning to create a fork of this branch, and add my Constantine fork as a submodule and keep integrating, once my PR gets merged, we can merge the fork to this branch, replace the Constantine fork submodule to Constantine master, and later merge this branch to nim-eth-verkle, once everything is done.

**Comment by advaita-saha** [comment]
> I can create a tag or you can target a specific commit, just in case I somehow break something.

For now we will be targeting a specific commit. Once things are stable, you can create a tag. I think that would be better

File: eth_verkle/verkle_types/types.nim
Comment (mratsim) [inline eth_verkle/verkle_types/types.nim]: This can be const

**Review by mratsim — COMMENTED** [review]
From the import, I think this should be in Constantine, is it an IPA implementation?

**Comment by advaita-saha** [comment]
> From the import, I think this should be in Constantine, is it an IPA implementation?

Yeah
For now to speed up the integration, we added the types file. Once the IPA is done, it will be merged into constantine & we can remove this types file from `nim-eth-verkle`. 
For now this will help us correct the implementations for the verkle trie

File: eth_verkle/verkle_types/types.nim
Comment (agnxsh) [reply eth_verkle/verkle_types/types.nim]: Resolving it

File: eth_verkle/tree/commitment.nim
Comment (mratsim) [inline eth_verkle/tree/commitment.nim]: ideally those imports are all reexported from constantine/eth_verkle_ipa

Comment (agnxsh) [reply eth_verkle/tree/commitment.nim]: i think this should do, as the API has been upstreamed
```nim
 ../../constantine/constantine/ethereum_verkle_trees
```

**Review by daniellm — APPROVED** [review]

**Pull request merged**
